File: im-mobile/lib/services/reaction_service.dart
import 'dart:convert';
import '../models/reaction.dart';
import 'api_service.dart';

class ReactionService {
  final ApiService _api = ApiService();

  Future<void> addReaction(String messageId, String emoji, {String type = 'EMOJI'}) async {
    await _api.post('/api/reactions/add', params: {
      'messageId': messageId,
      'userId': _api.currentUserId,
      'emoji': emoji,
      'type': type,
    });
  }

  Future<void> removeReaction(String messageId, String emoji) async {
    await _api.post('/api/reactions/remove', params: {
      'messageId': messageId,
      'userId': _api.currentUserId,
      'emoji': emoji,
    });
  }

  Future<List<ReactionWithUsers>> getReactions(String messageId) async {
    final resp = await _api.get('/api/reactions/message/$messageId');
    final List<dynamic> data = resp.data;
    return data.map((e) => ReactionWithUsers.fromJson(e)).toList();
  }

  Future<ReactionStats> getStats(String messageId) async {
    final resp = await _api.get('/api/reactions/stats/$messageId');
    return ReactionStats.fromJson(resp.data);
  }

  void subscribeToReactionUpdates(void Function(String, ReactionStats) callback) {
    _api.subscribe('reaction_update', (data) {
      callback(data['messageId'], ReactionStats.fromJson(data));
    });
  }
}

File: im-desktop/src/services/reaction_service.ts
import { apiClient, wsClient } from './client';

export interface ReactionStats {
  messageId: string;
  counts: Record<string, number>;
}

export interface ReactionWithUsers {
  emoji: string;
  count: number;
  userIds: string[];
}

class ReactionService {
  async addReaction(messageId: string, emoji: string, type = 'EMOJI'): Promise<void> {
    await apiClient.post('/api/reactions/add', null, {
      params: { messageId, userId: apiClient.userId, emoji, type }
    });
  }

  async removeReaction(messageId: string, emoji: string): Promise<void> {
    await apiClient.post('/api/reactions/remove', null, {
      params: { messageId, userId: apiClient.userId, emoji }
    });
  }

  async getReactions(messageId: string): Promise<ReactionWithUsers[]> {
    const res = await apiClient.get(`/api/reactions/message/${messageId}`);
    return res.data;
  }

  async getStats(messageId: string): Promise<ReactionStats> {
    const res = await apiClient.get(`/api/reactions/stats/${messageId}`);
    return res.data;
  }

  onReactionUpdate(callback: (messageId: string, stats: ReactionStats) => void): void {
    wsClient.on('reaction_update', (data: any) => {
      callback(data.messageId, { messageId: data.messageId, counts: data.stats });
    });
  }
}

export const reactionService = new ReactionService();
